src/interface/pipe.rs

```diff
@@ -89,60 +168,127 @@ impl EmulatedPipe {
+    ///
+    /// To learn more about pipes and the write syscall
+    /// [pipe(7)](https://man7.org/linux/man-pages/man7/pipe.7.html)
+    /// [write(2)](https://man7.org/linux/man-pages/man2/write.2.html)
     pub fn write_to_pipe(&self, ptr: *const u8, length: usize, nonblocking: bool) -> i32 {
-        let mut bytes_written = 0;
-
+        // convert the raw pointer into a slice to interface with the circular buffer
         let buf = unsafe {
             assert!(!ptr.is_null());
             slice::from_raw_parts(ptr, length)
         };
 
         let mut write_end = self.write_end.lock();
-
-        let pipe_space = write_end.remaining();
-        if nonblocking && (pipe_space == 0) {
-            return syscall_error(
-                Errno::EAGAIN,
-                "write",
-                "there is no data available right now, try again later",
-            );
-        }
+        let mut bytes_written = 0;
 
         while bytes_written < length {
             if self.get_read_ref() == 0 {
+                // we send EPIPE here since all read ends are closed
                 return syscall_error(Errno::EPIPE, "write", "broken pipe");
-            } // EPIPE, all read ends are closed
+            }
 
             let remaining = write_end.remaining();
 
-            if remaining == 0 {
-                interface::lind_yield(); //yield on a full pipe
-                continue;
-            }
-            // we write if the pipe is empty, otherwise we try to limit writes to 4096 bytes (unless whats leftover of this write is < 4096)
-            if remaining != self.size
-                && (length - bytes_written) > PAGE_SIZE
-                && remaining < PAGE_SIZE
-            {
-                continue;
+            // we loop until either more than a page of bytes is free in the pipe
+            // for why we wait for a free page of bytes, refer to the pipe man page about atomicity of writes or the pipe_write kernel implementation
+            if remaining < PAGE_SIZE {
+                if nonblocking {
+                    // for non-blocking if we have written a bit lets return how much we've written, otherwise we return EAGAIN
+                    if bytes_written > 0 {
+                        return bytes_written as i32;
+                    } else {
+                        return syscall_error(
+                            Errno::EAGAIN,
+                            "write",
+                            "there is no space available right now, try again later",
+                        );
+                    }
+                } else {
+                    // we yield here on a non-writable pipe to let other threads continue more quickly
+                    interface::lind_yield();
+                    continue;
+                }
             };
+
+            // lets read the minimum of the specified amount or whatever space we have
             let bytes_to_write = min(length, bytes_written as usize + remaining);
```

Comment:
Just want to point out that Linux is writing a maximum of one page at each iteration, instead of writing all bytes at the same time. Though I think this might not be an issue